Update 2020-10-02.md (#42)
rhodyprog4ds/BrownFall20: notes/2020-10-02.md

```{code-cell} ipython3
print(ca_dds_df)
```
we can do df.index to look at what the index is in our dataframe
```{code-cell} ipython3
ca_dds_df.index
```

> columns are tuple values. We can index to the first column as below. As we can see by running this code, this column has three names initially. 
> ```{code-cell} ipython3
> ca_dds_df.columns[0]
> ```

Reviewer: ```suggestion
Next, we can look at the columns, first we can index to the first column as below. 
```{code-cell} ipython3
ca_dds_df.columns[0]
```
As we can see by running this code, this column has three names initially. 
```

> By the following code we can assign the first column as the index of original dataframe. If we assign "False" to the "inplace", it will make the changes in a new dataframe.

Reviewer: ```suggestion
We can use this to set the first column as the index of original DataFrame, without having to copy that whole long name manually. The default of `set_index` method is to modify a copy of the DataFrame and return that because `inplace = False`. We want to change our existing DataFrame so we use `inplace=True`
```

The backticks will render as code in the built website.

```{code-cell} ipython3
ca_dds_df.set_index(ca_dds_df.columns[0],inplace=True)
```

> The updated dataframe looks like below. We still have the name of index inappropriate and aslo there are empty rows. 

Reviewer: ```suggestion
The updated dataframe looks like below. The name of index is still not informative and there are empty rows. 
```

```{code-cell} ipython3
ca_dds_df
```

> we can drop rows with "all" or "any",  "NaN" values. We would choose "all", since if we choose "any", we will loose many rows here and we do not want it.

Reviewer: ```suggestion
we can drop rows with rows or columns with `NaN` values in two ways: "all" or "any". The default is `any` which will drop a row if it has _any_ values that are `NaN`. We will use "all" instead, because we only want to drop the rows with no data in them.
```

Your description here is close, but not detailed.

```{code-cell} ipython3
ca_dds_df.dropna(how='all',inplace=True)
```

> We can view the index by the following code. Remember that this is not a regular column anymore since we assigned it to be our dataframe index.

Reviewer: ```suggestion
We can view the index by the following code. Remember that this is not a regular column anymore since we assigned it to be the index of our DataFrame.
```

```{code-cell} ipython3
ca_dds_df.index
```
Now lets rename the index to a more meaningful name as below.
```{code-cell} ipython3
ca_dds_df.index.rename('Age Cohort',inplace=True)
```
But we still have multi-row headers which should get corrected. We also want to have columns for "Age", "Race", "Gender", "Mean" and "Count".

> Let's flip the dataframe and unstack the index. For this purpose, we should unstack it in two levels of "Race" and "Gender".  After unstacking we want to flip it again and finally reset the index.
> ```{code-cell} ipython3
> ca_dds_df.T.unstack(level= [0,1]).T.reset_index()
> ```

Reviewer: For annotation, this can be re-split into multiple cells. Also add a bit more description of what we saw and look for at each step
````suggestion
Let's flip the dataframe
```{code-cell} ipython3
ca_dds_df.T
```
  unstack the index. For this purpose, we should unstack it in two levels of "Race" and "Gender".  
```{code-cell} ipython3
ca_dds_df.T.unstack(level= [0,1])
```
After unstacking we want to flip it again and finally reset the index.
```{code-cell} ipython3
ca_dds_df.T.unstack(level= [0,1]).T.reset_index()
```
````

> All these changes will be saved into a new dataframe named "ca_dds_clean".

Reviewer: `````suggestion
````{margin}
```{tip}
we worked without saving our changes to view what happened at each step. Now, we'll ave them to a new DataFrame since this is a big change and doing all of those operations again in place would be more lines than just assigning to a new variable
```
````
All these changes will be saved into a new dataframe named "ca_dds_clean".
`````

added a margin not with a bit more detail

```{code-cell} ipython3
ca_dds_clean = ca_dds_df.T.unstack(level= [0,1]).T.reset_index()
ca_dds_clean.head()
```

> Now let's change name of columns to appropriate names. 
> As we can see here, we can not change elements of an index. Instead we should reassign the whole column names.
> The reason is that index variable's type is type "index" not a " list", "tuple" or a "regular python based datatype". Index has special properties and python does not allow mutaple operations or assigning values to that. 
> ```{code-cell} ipython3
> ca_dds_clean.columns[1] = 'Race'
> ```

Reviewer: ```suggestion

Now let's change name of columns to be formatted better for Python. We want to rename the one in index 1 to race. 
```{code-cell} ipython3
ca_dds_clean.columns[1] = 'Race'
```
```suggestion
As we can see here, we can not change elements of an index. Instead we should reassign the whole column names.
The reason is that index variable's type is type `pandas` `Index` not a regular python based datatype.  Index has special properties and python does not allow mutable operations, such as assigning it a new value. 
```

```

> Column names are as below before changing them.

Reviewer: ```suggestion
```
that's not what happens in this line

```{code-cell} ipython3
type(ca_dds_clean.columns)
```

> The way we can rename index by using ".rename()" function and assigning a dictionary of old and new names to the column argument of this fuction.

Reviewer: ```suggestion
The way we can rename index by using ".rename()"  method with a dictionary that maps new names (as the values) to the current names (the keys of the dictionary). We can either tell `rename` what `axis` of the DataFrame to work on or pass our mapper to the `columns` keyword. 
```
